[API] Add -uri option to obtain a specific project

josephmyers opened the pull request:
When given a -uri argument, the Obtain command will open the GetCloneFromInternetDialog and clone a new repo from the URI. I opted to expand this command over creating a new one, as it seemed a less intrusive and breaking change.

The URI is stripped of everything but its scheme, hostname, and project name (first segment). Further, a simple dialog box is shown on launch, confirming the download. These features help with security.

The operation returns a CloneResult, just like the "any project" operation, and will presumably proceed to open the new project.

The current design requires an environment variable to be set containing the JWT password. This allows us to avoid passing it through CMD to this process. It's recommended to instantiate the env var with scope limited to the process.

The JWT is coupled with the "bearer" username and therefore required. Other user/pass combinations are not supported.


<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/flexbridge/388)
<!-- Reviewable:end -->

josephmyers committed:
Obtain command can now clone a specific project

When given a -uri argument, the Obtain command will open the GetCloneFromInternetDialog and clone a new repo from the URI. I opted to expand this command over creating a new one, as it seemed a less intrusive and breaking change.

The URI is stripped of everything but its scheme, hostname, and project name (first segment). Further, a simple dialog box is shown on launch, confirming the download. These features help with security.

The operation returns a CloneResult, just like the "any project" operation, and will presumably proceed to open the new project.

The current design requires an environment variable to be set containing the JWT password. This allows us to avoid passing it through CMD to this process. It's recommended to instantiate the env var with scope limited to the process.

The JWT is coupled with the "bearer" username and therefore required. Other user/pass combinations are not supported.

josephmyers committed:
Requiring the project name to be passed explicitly

This gives us more flexibility (less security) for changing URI structure, as it no longer parses the URI to determine where to clone from. The project name, at this point, just functions as the desired local folder name.

papeh reviewed (changes requested):
A more-descriptive title would be helpful; something like:
[API] Add -uri option to obtain a specific project

This should include a changelog entry and at least a minor version bump.

The commit comment should mention https://github.com/sillsdev/languageforge-lexbox/issues/242

What is JWT? Some comments in code about the use case for JWT and `bearer` are merited. Better yet, don't reference either, but allow both username and password to be specified by the client (or perhaps read from Chorus settings) so we aren't hard-coding a specific use case.

hahn-kev commented:
The JWT is effectively the password in this case and it is being specified by the client (it's coming from an environment variable), but since it's a password we don't want to pass it over command line args because that exposes it to any other application on the system. Good point about the user though, we might as well pass those all from the client side.

josephmyers committed:
Updated based on API changes

josephmyers committed:
Added -user option

And updated changelog

Part of https://github.com/sillsdev/languageforge-lexbox/issues/242

+semver:minor

josephmyers committed:
Missed a file

papeh reviewed (changes requested):
> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: all files reviewed, 4 unresolved discussions (waiting on @josephmyers)

___
*[`src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs` line 20 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-MZrO4G11lYEuHT-B:-Ng-MZrO4G11lYEuHT-C:b-8xv4n9) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs#L20)):*
> ```Smalltalk
> 	{
> // ReSharper disable InconsistentNaming
> 		internal const string u = "-u"; // userNameActual
> ```

Now would be a good time to update this comment:  // the username to use in the commit log
___
*[`src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs` line 32 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-NoaVEv2zPXOJBLwo:-Ng-NoaVEv2zPXOJBLwp:b-824dpa) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs#L32)):*
> ```Smalltalk
> 		internal const string locale = "-locale";
> 		internal const string uri = "-uri";
> 		internal const string project = "-project";
> ```

These options could use comments.

FYI, for `obtain`, `-p` is the projects directory. For `send_rec`, `-p` is the full path to the .fwdata file.

`-project` is acceptable, but a more descriptive name may be better.
___
*[`src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs` line 33 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-MjdABzojzyCG-qfs:-Ng-MjdABzojzyCG-qft:bivhw2b) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs#L33)):*
> ```Smalltalk
> 		internal const string uri = "-uri";
> 		internal const string project = "-project";
> 		internal const string user = "-user";
> ```

Adding a second `user` option is confusing. I'd prefer the credentials be passed together (parallel env vars, or a single `user:pass` var).
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 119 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-QNT-8M2Ofie4-HOV:-Ng-QNT-8M2Ofie4-HOW:b-s3ssvq) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L119)):*
> ```Smalltalk
> 		{
> 			var uri = new Uri(uriArg);
> 			var jwt = Environment.GetEnvironmentVariable("JWT");
> ```

I don't know what JWT is. I assume this TLA is specific to LanguageForge-LexBox. Since this is a shared library, we should use a clearer, more generic name.

Ideas: LD_PASS, LF_PASS, CHORUS_CREDENTIALS, FB_CREDS, or similar. Since we expect the variable's scope to be limited to the process tree, just plain CREDENTIALS should also work. (See comment above about passing the username the same way as the password).
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 122 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-VYHX9pe4giavmQ63:-Ng-VYHX9pe4giavmQ64:b-mhag45) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L122)):*
> ```Smalltalk
> 			var host = new Uri(uri.GetComponents(UriComponents.SchemeAndServer, UriFormat.SafeUnescaped));
> 
> 			if (MessageBox.Show($"Download {projectArg} from {host}?", "Confirm Download", MessageBoxButtons.YesNo) != DialogResult.Yes) return null;
> ```

Technically, `projectArg` is the local project name, which may or may not be the name on the server. Although the difference may not be worth the effort.

Optionally, this prompt could be moved to a new Chorus method `PromptAndStartClone(user, pass, path, proj, uri)`
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 124 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-ZDii9x9ECmfDh5Uy:-Ng-ZDii9x9ECmfDh5Uz:bbahpi3) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L124)):*
> ```Smalltalk
> 			if (MessageBox.Show($"Download {projectArg} from {host}?", "Confirm Download", MessageBoxButtons.YesNo) != DialogResult.Yes) return null;
> 
> 			var dialog = GetCloneFromInternetDialog.StartClone(userArg, jwt, _pathToRepository, projectArg, uri);
> ```

This line and following should all go into Chorus.


<!-- Sent from Reviewable.io -->

josephmyers committed:
Updates for API changes

josephmyers committed:
Moved username into env var with password.

josephmyers reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: 1 of 3 files reviewed, 4 unresolved discussions (waiting on @papeh)

___
*[`src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs` line 33 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-MjdABzojzyCG-qfs:-Ng2WPW78OuT6wteOtCO:b-896fix) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs#L33)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

Adding a second `user` option is confusing. I'd prefer the credentials be passed together (parallel env vars, or a single `user:pass` var).
</blockquote></details>

Done.
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 119 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-QNT-8M2Ofie4-HOV:-Ng2WR528Vo6gfzbOfec:b-896fix) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L119)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

I don't know what JWT is. I assume this TLA is specific to LanguageForge-LexBox. Since this is a shared library, we should use a clearer, more generic name.

Ideas: LD_PASS, LF_PASS, CHORUS_CREDENTIALS, FB_CREDS, or similar. Since we expect the variable's scope to be limited to the process tree, just plain CREDENTIALS should also work. (See comment above about passing the username the same way as the password).
</blockquote></details>

Done.
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 122 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-VYHX9pe4giavmQ63:-Ng2WUH_7WJM3cqChrwu:b-76b30b) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L122)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

Technically, `projectArg` is the local project name, which may or may not be the name on the server. Although the difference may not be worth the effort.

Optionally, this prompt could be moved to a new Chorus method `PromptAndStartClone(user, pass, path, proj, uri)`
</blockquote></details>

Hm, I'm either way. It's totally up to you!
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 124 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-ZDii9x9ECmfDh5Uy:-Ng2XesOAd_XDQ0aXYdy:b-896fix) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L124)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

This line and following should all go into Chorus.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

hahn-kev commented:
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 119 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-QNT-8M2Ofie4-HOV:-Ng3SIFu1WiOkfl2IWAx:b-5wy0rw) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L119)):*
<details><summary><i>Previously, josephmyers wrote…</i></summary><blockquote>

Done.
</blockquote></details>

jwt is a standard name, we could call it the LD_JWT but JWT is a full description of what it is. See https://jwt.io/ if you're not sure but it's a pretty standard term.


<!-- Sent from Reviewable.io -->

papeh reviewed (commented):
> Reviewed 1 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: 2 of 3 files reviewed, 3 unresolved discussions (waiting on @josephmyers)

___
*[`src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs` line 32 at r3](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng62iGvCKt-XxSZdhbN:-Ng62iGvCKt-XxSZdhbO:b4mu83j) ([raw file](https://github.com/sillsdev/flexbridge/blob/9ba9a5a400af3d241b507d98032046c4d438ffc0/src/TriboroughBridge-ChorusPlugin/CommandLineProcessor.cs#L32)):*
> ```Smalltalk
> 		internal const string locale = "-locale";
> 		internal const string uri = "-uri"; // Full project URI
> 		internal const string project = "-project"; // Project name
> ```

Project name to clone into
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 119 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-QNT-8M2Ofie4-HOV:-Ng5lnfK9HOmPghWpR_5:bnm4trj) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L119)):*
<details><summary><i>Previously, hahn-kev (Kevin Hahn) wrote…</i></summary><blockquote>

jwt is a standard name, we could call it the LD_JWT but JWT is a full description of what it is. See https://jwt.io/ if you're not sure but it's a pretty standard term.
</blockquote></details>

Thanks for clarifying. I still prefer the user:pass language, since it's consistent with Chorus. `CHORUS_CREDENTIALS` is good because it can go either way (user:pass or bearer:token).
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 122 at r2](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng-VYHX9pe4giavmQ63:-Ng62thS7ElkjoJo4-CW:b-4neolh) ([raw file](https://github.com/sillsdev/flexbridge/blob/1cec057e4dffb3cdc5d37ee8a3d10cf199fd0910/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L122)):*
<details><summary><i>Previously, josephmyers wrote…</i></summary><blockquote>

Hm, I'm either way. It's totally up to you!
</blockquote></details>

I think the prompt would be better in Chorus (PromptAndClone or PromptAndDoClone)
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 120 at r3](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng5mBOv98optmOB5_8G:-Ng5mBOv98optmOB5_8H:b-2r1m92) ([raw file](https://github.com/sillsdev/flexbridge/blob/9ba9a5a400af3d241b507d98032046c4d438ffc0/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L120)):*
> ```Smalltalk
> 			var userpass = Environment.GetEnvironmentVariable("CHORUS_CREDENTIALS");
> 			var user = userpass.Split(':')[0];
> 			var pass = userpass.Split(':')[1];
> ```

How likely is a token to contain ':'? (a username never should, so that part is fine)


<!-- Sent from Reviewable.io -->

hahn-kev reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: 2 of 3 files reviewed, 3 unresolved discussions (waiting on @josephmyers and @papeh)

___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 120 at r3](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng5mBOv98optmOB5_8G:-NgHk43587PvquFIDRWk:bg7nerw) ([raw file](https://github.com/sillsdev/flexbridge/blob/9ba9a5a400af3d241b507d98032046c4d438ffc0/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L120)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

How likely is a token to contain ':'? (a username never should, so that part is fine)
</blockquote></details>

JWT is encoded and will not include a ':' but a user password could quite easily, in that case we should just use `userpass.Split(':', 2)`, then any ':' in the password will be ignored by split.


<!-- Sent from Reviewable.io -->

papeh committed:
Move confirmation dialog to Chorus

Change-Id: Idc466778e41f5efa5536c68292188d16c7cdba06

papeh reviewed (commented):
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: all files reviewed, 1 unresolved discussion (waiting on @josephmyers)

___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 118 at r3](https://reviewable.io/reviews/sillsdev/flexbridge/388#-NgKgksCDL07RU0-AdMm:-NgKgksCDL07RU0-AdMn:b-2kve7r) ([raw file](https://github.com/sillsdev/flexbridge/blob/9ba9a5a400af3d241b507d98032046c4d438ffc0/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L118)):*
> ```Smalltalk
> 		{
> 			var uri = new Uri(uriArg);
> 			var userpass = Environment.GetEnvironmentVariable("CHORUS_CREDENTIALS");
> ```

We should guard against `null` or `!Contains(":")`
___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 120 at r3](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Ng5mBOv98optmOB5_8G:-NgKgfv01hIrLQDPYcdp:bb6un1j) ([raw file](https://github.com/sillsdev/flexbridge/blob/9ba9a5a400af3d241b507d98032046c4d438ffc0/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L120)):*
<details><summary><i>Previously, hahn-kev (Kevin Hahn) wrote…</i></summary><blockquote>

JWT is encoded and will not include a ':' but a user password could quite easily, in that case we should just use `userpass.Split(':', 2)`, then any ':' in the password will be ignored by split.
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

josephmyers committed:
PR comments

josephmyers reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: 1 of 3 files reviewed, 1 unresolved discussion (waiting on @papeh)

___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 118 at r3](https://reviewable.io/reviews/sillsdev/flexbridge/388#-NgKgksCDL07RU0-AdMm:-Ngbly6l5NBBbJ-V0pDX:b-896fix) ([raw file](https://github.com/sillsdev/flexbridge/blob/9ba9a5a400af3d241b507d98032046c4d438ffc0/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L118)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

We should guard against `null` or `!Contains(":")`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

josephmyers commented:
You're probably going to have a comment about my logic to detect if we should persist the user/pass. I am open to making it more configurable 👍

papeh reviewed (commented):
Short of adding another parameter, or a check box on the Chorus dialog to "Remember <bearer>'s credentials?", your persistence logic is fine. We'll see what @jasonleenaylor says. There should be a code comment that explains why and links to https://github.com/sillsdev/languageforge-lexbox/issues/242.

If Bear E. Robertson complains, we can apologize all over ourselves.

> Reviewed 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388#-:-NgdxZb683w83uQLy_f3:b9jf3ml)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @josephmyers)



<!-- Sent from Reviewable.io -->

josephmyers committed:
Added comment

josephmyers committed:
Updated to latest version of Chorus

papeh reviewed (commented):
> Reviewed 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: all files reviewed, 1 unresolved discussion (waiting on @josephmyers)

___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 122 at r6](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Nh7C6G96jvNyfUBBRNk:-Nh7C6G96jvNyfUBBRNl:ba5m0x0) ([raw file](https://github.com/sillsdev/flexbridge/blob/4cfe29895764e408fe25fb511fdc78fe539b125b/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L122)):*
> ```Smalltalk
> 			var user = split[0];
> 			var pass = split[1];
> 			//Accepted practice for authenticating with JWT (https://github.com/sillsdev/languageforge-lexbox/issues/242)
> ```

comment should be more verbose. The audience may be a new FLEx developer who's never heard of LexBox. The developer may wonder:

Why do we care if it's JWT? What do we do differently? Why?


<!-- Sent from Reviewable.io -->

josephmyers committed:
Another comment

josephmyers committed:
Fixed typo

josephmyers commented:
From my side, this is more or less done, but I'd like you guys to see the FieldWorks side before we sign off. I'm currently waiting on permissions to push to FieldWorks (I emailed the gerrit account)

josephmyers commented:
https://gerrit.lsdev.sil.org/c/FieldWorks/+/9549

josephmyers commented:
Please note that the FieldWorks change I was able to push is not ready for review. But it should provide good context for this PR.

papeh reviewed (commented):
> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @josephmyers)



<!-- Sent from Reviewable.io -->

josephmyers reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @josephmyers)

___
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 122 at r6](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Nh7C6G96jvNyfUBBRNk:-NheKifSCndqKohWGWJr:bb6un1j) ([raw file](https://github.com/sillsdev/flexbridge/blob/4cfe29895764e408fe25fb511fdc78fe539b125b/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L122)):*
<details><summary><i>Previously, papeh wrote…</i></summary><blockquote>

comment should be more verbose. The audience may be a new FLEx developer who's never heard of LexBox. The developer may wonder:

Why do we care if it's JWT? What do we do differently? Why?
</blockquote></details>

Done


<!-- Sent from Reviewable.io -->

papeh reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Barring any objections from @jasonleenaylor, this can be squashed and merged.

> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388#-:-Nhg3EbqDeA6WvXNoCyv:b-32j1g0)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @josephmyers)



<!-- Sent from Reviewable.io -->

josephmyers committed:
Add support for opening pre-existing projects with custom names

If a -repositoryIdentifier (aka commit) is supplied from FLEx, we now look through the existing projects for a matching ID. If we find a match, progress through the clone as normal, indicating that a clone happened.

In coordination with this change, the FLEX obtain strategy now checks for previously existing .fwdata files in the indicated cloneLocation. If it finds such, it does not attempt to "put humpty together again."

Finally, in this case, it will proceed to "tell FLEx about it" in the respective method.

josephmyers commented:
Sorry, @papeh, I've still been having to make changes to this in response to FieldWorks changes. In particular, we needed a way to support non-default project names. I thought we would be able to do this from FieldWorks. I do think this repo is close, though

josephmyers commented:
Can anyone rerun this build, please?

papeh commented:
>Can anyone rerun this build, please?

Will do. If you log in to AppVeyor with GitHub authentication, you should automatically have the appropriate permissions.

josephmyers commented:
> > Can anyone rerun this build, please?
> 
> Will do. If you log in to AppVeyor with GitHub authentication, you should automatically have the appropriate permissions.

Ok. I had done that last week and wasn't able to see where to do so. Maybe I just don't know their UI well. I've logged out and back in and am still unable to find how to do it. Is it possible I'm not on the project? I can see individual builds, using links from this page in GH, but in AppVeyor my Projects page is empty.

josephmyers commented:
@jasonleenaylor This guy is ready for review

jasonleenaylor commented:
*[`src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs` line 148 at r8](https://reviewable.io/reviews/sillsdev/flexbridge/388#-Nifey7t1M-lkwuBsVaq:-Nifey7t1M-lkwuBsVar:bpwwdfd) ([raw file](https://github.com/sillsdev/flexbridge/blob/3955d3c9a4108d50661dec94ce22e5733e350de3/src/TriboroughBridge-ChorusPlugin/Infrastructure/ActionHandlers/ObtainAnyProjectActionHandler.cs#L148)):*
> ```Smalltalk
> 		{
> 			var credentials = Environment.GetEnvironmentVariable("CHORUS_CREDENTIALS");
> 			if (credentials == null || !credentials.Contains(":")) return new CloneResult(null, CloneStatus.NotCreated);
> ```

please put the return on a new line.


<!-- Sent from Reviewable.io -->

jasonleenaylor reviewed (commented):
> Reviewed 2 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: 4 of 5 files reviewed, 1 unresolved discussion (waiting on @hahn-kev, @josephmyers, and @papeh)



<!-- Sent from Reviewable.io -->

hahn-kev reviewed (commented):
> Reviewed 1 of 3 files at r2, 1 of 2 files at r6, 3 of 3 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/flexbridge/388)* status: all files reviewed, 1 unresolved discussion (waiting on @josephmyers)



<!-- Sent from Reviewable.io -->

josephmyers committed:
PR comment
